### PR TITLE
Improve `Debug` impl of `time::Duration`

### PR DESCRIPTION
**Pull request by LukasKalbertodt**
Hi there!

For a long time now, I was getting annoyed by the derived `Debug` impl of `Duration`. Usually, I use `Duration` to either do quick'n'dirty benchmarking or measuring the time of some operation in general. The output of the derived Debug impl is hard to parse for humans: is { secs: 0, nanos: 968360102 } or { secs: 0, nanos 98507324 } longer? 

So after running into the annoyance several times (sometimes building my own function to print the Duration properly), I decided to tackle this. Now the output looks like this:

```
Duration::new(1, 0)                 => 1s
Duration::new(1, 1)                 => 1.000000001s
Duration::new(1, 300)               => 1.0000003s
Duration::new(1, 4000)              => 1.000004s
Duration::new(1, 600000)            => 1.0006s
Duration::new(1, 7000000)           => 1.007s
Duration::new(0, 0)                 => 0ns
Duration::new(0, 1)                 => 1ns
Duration::new(0, 300)               => 300ns
Duration::new(0, 4001)              => 4.001µs
Duration::new(0, 600300)            => 600.3µs
Duration::new(0, 7000000)           => 7ms
```

Note that I implemented the formatting manually and didn't use floats. No information is "lost" when printing. So `Duration::new(123_456_789_000, 900_000_001)` prints as `123456789000.900000001s`.

~~This is not yet finished~~, but I wanted to open the PR now already in order to get some feedback (maybe everyone likes the derived impl).

### Still ToDo:

- [x] Respect precision ~~and width~~ parameter of the formatter (see [this comment](https://github.com/rust-lang/rust/pull/50364#issuecomment-386107107))

### Alternatives/Decisions

- Should large durations displayed in minutes, hours, days, ...? For now, I decided not to because the current formatting is close the how a `Duration` is stored. From this new `Debug` output, you can still easily see what the values of `secs` and `nanos` are. A formatting like `3h 27m 12s 9ms` might be more appropriate for a `Display` impl?
- Should this rather be a `Display` impl and should `Debug` be derived? Maybe this formatting is too fancy for `Debug`? In my opinion it's not and, as already mentioned, from the current format one can still very easily determine the values for `secs` and `nanos`.
- Whitespace between the number and the unit?

### Notes for reviewers

- ~~The combined diff sucks. Rather review both commits individually.~~
- ~~In the unit test, I am building my own type implementing `fmt::Write` to test the output. Maybe there is already something like that which I can use?~~
- My `Debug` impl block is marked as `#[stable(...)]`... but that's fine since the derived Debug impl was stable already, right?

---

~~Apart from the main change, I moved all `time` unit tests into the `tests` directory. All other `libcore` tests are there, so I guess it was simply an oversight. Prior to this change, the `time` tests weren't run, so I guess this is kind of a bug fix. If my `Debug` impl is rejected, I can of course just send the fix as PR.~~ (this was already merged in #50466)

### CODE REVIEW

**Comment by LukasKalbertodt** [comment]
Failure caused by warnings in some of the tests I moved. So yeah, apparently those tests weren't executed in quite some time...

Hopefully those tests are fixed now.

**Comment by LukasKalbertodt** [comment]
I checked the different formatting options that this impl could respect, but I concluded that most are not worth it. When was the last time you specified any formatting options while using `?` for debug output? I think I never did...

- The whole fill/alignment/padding is pretty hard to implement (especially since this is in `core` and can't allocate) and I doubt that anyone would use that anyway. Additionally, even the `fmt` docs say: "Note that alignment may not be implemented by some types. A good way to ensure padding is applied is to format your input, then use this resulting string to pad your output."
- The alternate flag `#` is not really applicable here: it doesn't make sense to switch to pretty printing again
- The precision is a bit problematic. That's because the current impl switches between units depending on the value. What should `{:.2?}` mean? Always show two digits after the decimal point (`2.14s` but also `930.01µs`)? Or: show no values less than 1/100th of a second. I don't like either of those, but maybe the first makes more sense. I can implement it, if anyone thinks it's useful.

So, I just implemented the `+` flag now and will remove the "WIP" status as I think this PR is ready and needs feedback to continue.

**Comment by LukasKalbertodt** [comment]
- Rebased
- The code is a bit shorter now
- Respect the precision parameter of the formatter (I changed my mind and think it does make sense...)
- I noticed that I can use `format!()` in tests, so I removed my dirty hack from before
- I added more unit tests and split test cases in multiple test functions

Should be ready now ^_^

**Comment by LukasKalbertodt** [comment]
Summary: I don't quite care whether or not this becomes a `Debug` or `Display` impl. However, I think a `Display` impl could lead to long discussions about "the right default way to format". And I think my impl is not "incorrect" for the `Debug` trait and is fine the way it is.

---

My reasoning for implementing `Debug` was:

I assumed the `Display` trait is implemented only if there is one clear default way to represent that value as a string. For duration, I don't think everyone would agree on one default way (e.g. "should we print hours and minutes?"). The `fmt` docs state:
  
> `fmt::Display` implementations assert that the type can be faithfully represented as a UTF-8 string at all times. It is not expected that all types implement the `Display` trait.

> `Display` is similar to `Debug`, but `Display` is for user-facing output, and so cannot be derived.
  
This requirement doesn't sound too strict. However, it's still not clear what "user-facing output" for `Duration` should look like.

About `Debug`, the docs say "Output will typically represent the internal state as faithfully as possible." Note that it says "faithfully" but not "explicitly". In my opinion, my implementation does represent the internals faithfully, because it never omits information. Yes, figuring out the number of nano seconds requires some thinking, but not a lot (everything is just a base 10 off).

There are a number of other `Debug` impls which are also neither "easy" nor "explicit":
- `str`: some chars are escaped, but absolutely not all. So one could also argue that the programmer want's to see the raw bytes for debugging purposes. Also: length is now implicit.
- Collections, such as `Vec<T>`: Only the elements are printed, which means:
  - The length of the collection is implicit (you'd have to count)
  - The capacity of the collection is completely omitted

So `Debug` impls may in fact be non-trivial, making explicit information implicit, and apparently even hiding information if it's not important for most users.

**Comment by Stargateur** [comment]
You said 

> The output of the derived Debug impl is hard to parse for humans: is { secs: 0, nanos: 968360102 } or { secs: 0, nanos 98507324 } longer?

I could answer you with your own sentence (just need to inverse it):

>  Yes, figuring out the number of nano seconds requires some thinking, but not a lot (everything is just a base 10 off).

---

> Output will typically represent the internal state as faithfully as possible.

I agree with you about "faithfully" but the sentence clearly say "as possible". I think it's possible to be more accurate about the internal state of Duration than your new implementation as I show it in my previous exemple `1s (1s, 0n)`.

And your exemple with `Collection` and `str`, are a very strong argument but maybe they are wrong to do that ? I'm not a experimented rust user, so maybe this has been discute before, but in my opinion, I don't see why `Collection` and `str` don't print their length and their capacity.

> I assumed the Display trait is implemented only if there is one clear default way to represent that value as a string.
> And as a user: how often are you really interested in the exact values of secs and nanos in the explicit representation? I never was...

I really like your output as a user of `Duration`, this is why I propose `Display` because I think `Duration` is only use and read by the programmer so `Display` a `Duration` could be considerate as a the programmer user ? If I want something standard and the represent time that I will print to a final user (anyone), I would use `Chrono`. And as the documentation of `Duration` clearly state, "Each Duration is composed of a whole number of seconds and a fractional part represented in nanoseconds." so I think it's logical to print it like you do for a `Display` trait, no need for minute or hour. And maybe we could just agree about one way to represent `Duration` and starting to call it the standard way ;) ?

Your new implementation look nice enough for me to be a good candidate to `Display` it. But again all of this is just my little opinion, I would be happy even with your current proposition because I think it's a good improvement.

**Comment by KodrAus** [comment]
I'll dig into the source shortly, but just thought I'd leave a note now to say I'm in favour of this clearer `Debug` implementation. Representing values as pseudo structures in `Debug` is a good default that can always be derived, but when we can do better I think we should :+1:

File: src/libcore/tests/time.rs [hunk 1]
```diff
@@ -122,3 +122,115 @@ fn checked_div() {
...
+    // This might seem inconsistent with the other units, but printing
+    // fractional digits for nano seconds would imply more precision than is
+    // actually stored.
+    assert_eq!(format!("{:.2?}", Duration::new(0, 0)), "0ns");
+    assert_eq!(format!("{:.2?}", Duration::new(0, 123)), "123ns");
+
+    assert_eq!(format!("{:.2?}", Duration::new(0, 1_000)), "1.00µs");
+    assert_eq!(format!("{:.2?}", Duration::new(0, 7_001)), "7.00µs");
+    assert_eq!(format!("{:.2?}", Duration::new(0, 7_100)), "7.10µs");
+    assert_eq!(format!("{:.2?}", Duration::new(0, 1_999)), "1.99µs");
```
Comment (KodrAus) [inline src/libcore/tests/time.rs]: Hmm, the precision here is a bit inconsistent with floating points, that will also perform rounding through its unholy rendering machinery. For instance:

```rust
assert_eq!("2.00", format!("{:.2}", 1.995f32));
```

We might want to consider rounding here too.

Comment (LukasKalbertodt) [reply src/libcore/tests/time.rs]: Good point! I just pushed a commit which adds correct rounding like floats do. I also pushed a commit which fixes the behavior for a precision > 9.

**Comment by pietroalbini** [comment]
Can we get a crater run for this? Last time we changed the debug output there were some regressions (some crates assumes the debug output is guaranteed not to change, it would be nice to alert the authors).

**Comment by pietroalbini** [comment]
cc @rust-lang/infra

**Comment by pietroalbini** [comment]
Hi myself and @KodrAus! Crater results are at: http://cargobomb-reports.s3.amazonaws.com/pr-50364/index.html. 'Blacklisted' crates (spurious failures etc) can be found [here](https://github.com/rust-lang-nursery/crater/blob/master/config.toml). If you see any spurious failures not on the list, please make a PR against that file.

(interested observers: Crater is a tool for testing the impact of changes on the crates.io ecosystem. You can find out more at the [repo](https://github.com/rust-lang-nursery/crater/) if you're curious)

**Comment by LukasKalbertodt** [comment]
I quickly scanned the crater results and none of the 7 regressions look like they are caused by this change.

**Comment by KodrAus** [comment]
This looks good to me!

@bors r+

**Comment by KodrAus** [comment]
Thanks again for working on this @LukasKalbertodt!

**Comment by pietroalbini** [comment]
Failure to resove dns names.

@bors retry

**Comment by scottmcm** [comment]
Thanks for including precision!  For a real duration, I'd always rather `1.00s` over `1.000000001s`.

**Comment by uberjay** [comment]
Thank you for this! It's always nice when updating my compiler gives me prettier debug logging for free!
